runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp

```diff
@@ -411,19 +415,36 @@ lstore <offset>
     lconst ~J9_SUN_FIELD_OFFSET_MASK
 end Block_D
 
+start Block_G
+xcall atomic method helper
+  aladd
+    aload <object>
+    lload <offset>
+  xload value
+xstore
+  ==>xcall
+goto --> block_H
+end Block_G
+
 start Block_E
 xcall atomic method helper
   aladd
     aload <object>
     lload <offset>
   xload value
+xstore
+  ==>xcall
 end Block_E
```

Comment:
Is it possible to show clearly the "before" and "after" IL trees using this form of pseudocode ? This would make it easier to imagine what the issue was and how exactly it was fixed.

Reply:
It would also be good to have stated clearly what the copying of the call to the atomic method helper achieves exactly.

Reply:
Attach the diagram without the fix and the diagram with the fix here 
[without_the_fix.txt](https://github.com/eclipse-openj9/openj9/files/10143661/without_the_fix.txt)
[with_the_fix.txt](https://github.com/eclipse-openj9/openj9/files/10143639/with_the_fix.txt)

Diffing these two files will show where the changes are. Regarding to the temp sharing, without the fix, temp `object-1` is reused everywhere when a temp is needed. With the fix, temp `object-2` is introduced in `Block_D`. 

<img width="1228" alt="Screen Shot 2022-12-02 at 2 07 50 PM" src="https://user-images.githubusercontent.com/45463964/205367580-c4489994-c4bc-4ec4-8614-6d35473883fe.png">

<img width="725" alt="Screen Shot 2022-12-02 at 2 00 20 PM" src="https://user-images.githubusercontent.com/45463964/205366965-96159c52-cb03-401c-83d4-fa74e1889b43.png">

Reply:
>It would also be good to have stated clearly what the copying of the call to the atomic method helper achieves exactly.

Not sure I follow, which part of the code you are referring here?

Reply:
Sorry, I meant this part ```                                |                        |
                       calculate address         calculate address
                        for static field         for instance field
                                |                and absolute address
                                |                        |
                                |                        |
                                |                        |
                  xcall atomic method helper      xcall atomic method helper```

Reply:
Just double checking I understand it correctly: Do you mean adding comments like below to each of the helper call? 

```
                                |                          |
                       calculate address [D]     calculate address
                        for static field         for instance field
                                |                and absolute address
                                |                          |
                  xcall atomic method helper [G]     xcall atomic method helper [E]
                     (loads from static field)     (loads from instance field or absolute address)
                                |                          |
                                +-----------+--------------+
                                            |
                                            v
                             program after the original call [H]
```

Reply:
The bug described is something related to a temp being reused incorrectly with collected reference property.

That comment that you show talks about have two paths where we call that xcall atomic method helper instead of one path originally. Is this related to the temp reuse fix in some way ? If so, how ? And if not, please explain why that change was needed.

Reply:
Using the diagram in the comment to describe the issue here: 
The issue is that the temp `object-1` is reused for both collected reference and non collected reference (`ramStatics`) in `block_E`. The fix is to have `block_E` dedicated to load from the collected reference `object-1`, and add a new block `block_G`  that also calls atomic helper call but loads from another temp `object-2` which stores `ramStatics`. This temp `object-2` is non collected. 

I'll update the comments as below to clear state the difference on the references
 
```
                       calculate address [D]       calculate address
                        for static field           for instance field
                      (non collected reference)    and absolute address
                                |                 (collected reference)
                                |                          |
                  xcall atomic method helper [G]     xcall atomic method helper [E]
                                |                          |
                                +-----------+--------------+
                                            |
                                            v
                             program after the original call [H]
```

Reply:
Addressed in 072048a